File: solutions/129.sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
const sumNumbers = (root) => {
  if (!root) return 0
  let sum = 0

  const walk = (node, temp) => {
    if (!node) return

    if (!node.left && !node.right) {
      // leaf
      sum += (Number(temp + node.val))
    }

    if (node.left) walk(node.left, temp + node.val)
    if (node.right) walk(node.right, temp + node.val)
  }

  walk(root, '')

  return sum
}

File: solutions/129.sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        ret = 0
        if not root:
            return ret

        def traverse(node, temp):
            nonlocal ret
            if not node.left and not node.right:
                ret += temp * 10 + node.val

            if node.left:
                traverse(node.left, temp * 10 + node.val)

            if node.right:
                traverse(node.right, temp * 10 + node.val)

        traverse(root, 0)

        return ret
